src/core/json/path.cc

```diff
@@ -0,0 +1,223 @@
+            return nullptr;  // exhausted
+          }
+        }
+        break;
+      }
+      case SegmentType::INDEX: {
+        unsigned index = segment.index();
+        if (obj->is_array()) {
+          if (index >= obj->size()) {
+            return make_unexpected(OUT_OF_BOUNDS);  // exhausted
+          }
+          auto it = obj->array_range().cbegin() + index;
+          state = it;
+          return &*it;
+        }
+        break;
+      }
+
+      case SegmentType::WILDCARD: {
+        if (obj->is_object()) {
+          jsoncons::range rng = obj->object_range();
+          if (rng.cbegin() == rng.cend()) {
+            return nullptr;
+          }
+          auto it = rng.begin();
+          state = it;
+          return &it->value();
+        }
+
+        if (obj->is_array()) {
+          jsoncons::range rng = obj->array_range();
+          if (rng.cbegin() == rng.cend()) {
+            return nullptr;
+          }
+          state = rng.cbegin();
+          return &*rng.cbegin();
+        }
+        break;
+      }
+    }  // end switch
+
+    return make_unexpected(MISMATCH);
+  }  // end monostate
+
+  // If segment is set to sa single value, we can't advance.
+  if (segment.type() != SegmentType::WILDCARD) {
+    return nullptr;  // exhausted
+  }
+
+  // Advance for WILDCARD.
+  if (std::holds_alternative<JsonType::const_object_iterator>(state)) {
+    auto it = std::get<JsonType::const_object_iterator>(state);
+    ++it;
+    if (it == obj->object_range().end()) {
+      return nullptr;
+    }
+
+    state = it;
+    return &it->value();
+  }
+
+  CHECK(std::holds_alternative<JsonType::const_array_iterator>(state));
+  auto it = std::get<JsonType::const_array_iterator>(state);
+  ++it;
+  if (it == obj->array_range().end()) {
+    return nullptr;
+  }
+
+  state = it;
+  return &*it;
+}
+
+void Dfs::Traverse(absl::Span<const PathSegment> path, const JsonType& root, const Cb& callback) {
+  DCHECK(!path.empty());
+  if (path.size() == 1) {
+    PerformStep(path[0], root, callback);
+    return;
+  }
+
+  vector<Item> stack;
+  stack.emplace_back(&root);
+
+  do {
+    unsigned depth = stack.size();
+    // init or advance the current object
+    AdvanceResult res = stack.back().Advance(path[depth - 1]);
+    if (res && *res != nullptr) {
+      const JsonType* next = *res;
+      DVLOG(2) << "Handling now " << next->to_string();
+
+      if (depth + 1 < path.size()) {
+        stack.emplace_back(next);
+      } else {
+        // terminal step
+        // TODO: to take into account MatchStatus
+        // for `json.set foo $.a[10]` or for `json.set foo $.*.b`
+        PerformStep(path[depth], *next, callback);
+      }
+    } else {
+      stack.pop_back();
+    }
+  } while (!stack.empty());
+}
+
+auto Dfs::PerformStep(const PathSegment& segment, const JsonType& node,
+                      const function<void(const JsonType&)>& callback)
+    -> nonstd::expected<void, MatchStatus> {
+  switch (segment.type()) {
```

Comment:
Please consider using Visitor pattern with segment instead of switch (segment.type()).

Reply:
done

Reply:
segment is not a variant. I prefer keeping the switch, I think it's more straightforward.